lib/teleterm/clusters/cluster_access_requests.go

```diff
@@ -0,0 +1,262 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusters
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport/api/types"
+	apiutils "github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/services"
+	api "github.com/gravitational/teleport/lib/teleterm/api/protogen/golang/v1"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+
+	"github.com/gravitational/trace"
+)
+
+type AccessRequest struct {
+	URI uri.ResourceURI
+	types.AccessRequest
+}
+
+// Returns all access requests available to the user
+func (c *Cluster) GetAccessRequests(ctx context.Context, req *api.GetAccessRequestsRequest) ([]AccessRequest, error) {
+	var (
+		requests    []types.AccessRequest
+		authClient  auth.ClientI
+		proxyClient *client.ProxyClient
+		err         error
+	)
+	err = addMetadataToRetryableError(ctx, func() error {
+		proxyClient, err = c.clusterClient.ConnectToProxy(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer proxyClient.Close()
+
+		authClient, err = proxyClient.ConnectToCluster(ctx, c.clusterClient.SiteName)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer authClient.Close()
+
+		requests, err = authClient.GetAccessRequests(ctx, types.AccessRequestFilter{
+			ID:    req.Id,
+			State: types.RequestState(req.State),
+			User:  req.User,
+		})
+
+		return nil
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	results := []AccessRequest{}
+	for _, req := range requests {
+		results = append(results, AccessRequest{
+			URI:           c.URI.AppendAccessRequest(req.GetName()),
+			AccessRequest: req,
+		})
+	}
+
+	return results, nil
+}
+
+// Creates an access request
+func (c *Cluster) CreateAccessRequest(ctx context.Context, req *api.CreateAccessRequestRequest) (*AccessRequest, error) {
+	var (
+		err         error
+		authClient  auth.ClientI
+		proxyClient *client.ProxyClient
+		request     types.AccessRequest
+	)
+
+	resourceIDs := make([]types.ResourceID, 0, len(req.ResourceIds))
+	for _, resource := range req.ResourceIds {
+		resourceIDs = append(resourceIDs, types.ResourceID{
+			ClusterName: resource.ClusterName,
+			Name:        resource.Name,
+			Kind:        resource.Kind,
+		})
+	}
+
+	err = addMetadataToRetryableError(ctx, func() error {
+		proxyClient, err = c.clusterClient.ConnectToProxy(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer proxyClient.Close()
+
+		authClient, err = proxyClient.ConnectToCluster(ctx, c.clusterClient.SiteName)
```

Comment:
Currently access requests are always created/reviewed/assumed in the root cluster, and we rely on the cluster role mapping for access to leaf clusters. I'm not sure if this will always be the root cluster here, just want to make sure you're aware. Have you tested with a trusted cluster setup, and with access requests created from tsh/web ui and reviewed in connect?

Reply:
Yes, in Connect we pass the rootClusterUri during creating/reviewing/assuming. The local cluster is used only when fetching request-able resources essentially.

Reply:
I think changing the name of the field is better as it flows through to the frontend as well with `setRootClusterUri`